pkg/cloudprovider/aws.go

```diff
@@ -1,37 +1,338 @@
 package cloudprovider
 
 import (
+	"errors"
+	"fmt"
 	"net"
 
+	awsapi "github.com/aws/aws-sdk-go/aws"
+	awscredentials "github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
 	corev1 "k8s.io/api/core/v1"
+	utilnet "k8s.io/utils/net"
 )
 
 const (
 	aws = "aws"
 )
 
-// AWS implements the API wrapper for talking
-// to the AWS cloud API
+var NoNetworkInterfaceError error = errors.New("no retrievable network interface")
+
+var (
+	awsIPv4FilterKey = "network-interface.addresses.private-ip-address"
+	awsIPv6FilterKey = "network-interface.ipv6-addresses.ipv6-address"
+)
+
+// AWS implements the API wrapper for talking to the AWS cloud API
 type AWS struct {
-	cloud CloudProvider
+	CloudProvider
+	region string
+	client *ec2.EC2
+}
+
+// AWSWaitInput is the required input for the AWS EC2 Wait API call (WaitUntilInstanceRunning).
+// Unfortunately the API only handles equality assertion: so on delete we can't
+// specify and assert that the IP which is being removed is completely removed,
+// we are forced to do the inverse, i.e: assert that all IPs except the IP being
+// removed are there (which should anyways be the case). Hence we might risk
```

Comment:
That means the operation may fail if someone else modifies IPs at the same time we do, because the returned list will have either more or fewer IPs than we were expecting, right?

I don't know anything about the AWS APIs, but it seems weird that there is no better way to do this...